topics/data-science/tutorials/python-functions/tutorial.md

```diff
@@ -0,0 +1,300 @@
+
+objectives:
+- Understand the structure of a "function" in order to be able to construct their own functions and predict which functions will not work.
+
+time_estimation:  30M
+key_points:
+- Functions are foundational in Python
+- Everything you do will require functions
+
+subtopic: python-modular
+contributors:
+  - hexylena
+  - carpentries
+
+priority: 1
+notebook:
+  language: python
+
+abbreviations:
+  DRY: Don't Repeat Yourself
+---
+
+Functions are the basic unit of all work in Python! Absolutely everything you do uses functions. Conceptually, functions are super simple. Just like in maths, they take an input, do some transformation, and return an output. As an example, `f(x) = x + 2` is a function that calculates whatever value you request, plus two. But functions are foundational, so, you should understand them well before moving on.
+
+
+> ### Agenda
+>
+> In this tutorial, we will cover:
+>
+> 1. TOC
+> {:toc}
+>
+{: .agenda}
+
+
+## What is a Function
+
+Functions are a way to re-use some computation you want to do, multiple times. If you don't have a function, you need to re-write the calculation every time, so we use functions to collect those statements and make them easy to re-run. Additionally they let us "parameterise" some computation. Instead of computing the same value every time, we can template it out, we can re-run the computation with new inputs, and see new results.
+
+> > ### {% icon code-in %} Math
+> > ```
+> > # Define our function
+> > f(x) = 3 * x
+> > # Compute some value
+> > f(3) # is 9
+> > ```
+> {: .code-in}
+>
+> > ### {% icon code-out %} Python
+> > ```
+> > # Define our function
+> > def f(x):
+> >    return 3 * x
+> > # Compute some value
+> > f(3)
+> > ```
+> {: .code-out}
+>
+{: .code-2col}
+
+We've talked about mathematical functions before, but now we'll talk about more programing-related functions
+
+
+## Create Functions
+
+Human beings can only keep a few items in working memory at a time. Breaking down larger/more complicated pieces of code in functions helps in understanding and using it. A function can be re-used. Write one time, use many times. (Known as staying {DRY} in the industry.)
+
+Knowing what Americans mean when they talk about temperatures and weather can be difficult, but we can wrap the temperature conversion calculation ($$^{\circ}\text{C} = (^{\circ}\text{F} - 32) * \dfrac{5}{9}$$) up as a function that we can easily re-use.
+
+```python
+def fahr_to_celsius(temp):
+    return ((temp - 32) * (5/9))
+```
+
+![The above function fahr to celsius is shown except annotated. def is labelled "def statement", fahr_to_celsius is noted as the function name. Inside parentheses is temp and an arrow shows it is called parameter names. The next line which is indented is annotated as the function body which has a return statement and the calculation from above.](../../images/python-basics/Figure7_Functions.png)
```

Comment:
is this image a deadlink. 
It doesn't seem to work on the cocalc site

Reply:
Ahhh fixed.